src/python/pants/engine/legacy/parser.py

```diff
@@ -152,7 +157,47 @@ def check_for_deprecated_globs_usage(token: str, filepath: str, lineno: int) ->
                 hint=warning,
             )
 
-    def parse(self, filepath: str, filecontent: bytes):
+    @staticmethod
+    def _expose_symbols_from_loaded_files(
+        load_statements: Iterable[LoadStatementWithContent], filepath: str
+    ) -> Dict:
+        """Expose symbols from loaded files.
+
+        To achieve that, we exec() the contents of the imported file, which puts all defined symbols
+        in `locals()`. We then copy `locals()` and filter it by the symbols the build file
+        explicitly imports. `locals()` will be emptied at the end of each iteration of the outer
+        `for` loop, so there is no chance of contamination.
+        """
+        extra_symbols: Dict = {}
+        for statement in load_statements:
+            content = statement.content.content.decode()
+            symbols_to_expose = statement.symbols_to_expose
+            exec(content)
+            local_symbols = copy(
+                locals()
+            )  # We copy the locals so that they don't get lost in the for loop
+            for symbol in symbols_to_expose:
+                if symbol in local_symbols:
+                    if symbol not in extra_symbols:
+                        extra_symbols[symbol] = local_symbols[symbol]
+                    else:
+                        raise ParseError(
+                            f"Trying to import symbol {symbol} from file {statement.path} into BUILD file {filepath}: ",
+                            f"Which has already been loaded from another file.",
+                        )
+                else:
+                    raise ParseError(
+                        f"Trying to import symbol {symbol} from file {statement.path} into BUILD file {filepath}: ",
+                        f"Symbol is not exported by loaded file.",
```

Comment:
Do we intend to handle the case where the symbol is defined in the current BUILD file?
E.g.
```
diff --git a/examples/src/scala/org/pantsbuild/example/build_file_macros/BUILD b/examples/src/scala/org/pantsbuild/example/build_file_macros/BUILD
index 79fcb9c..ea0d08e 100644
--- a/examples/src/scala/org/pantsbuild/example/build_file_macros/BUILD
+++ b/examples/src/scala/org/pantsbuild/example/build_file_macros/BUILD
@@ -1,3 +1,5 @@
+dep_on_welcome = 123
+
 load("//examples/src/scala/org/pantsbuild/example/build_file_macros/BUILD_MACROS", "dep_on_welcome")
 load("examples/src/scala/org/pantsbuild/example/build_file_macros/BUILD_MACROS_2", "source_exe_scala")
```

```
$ ./pants list examples/src/scala/org/pantsbuild/example/build_file_macros/:
            pants.base.exceptions.MappingError: Failed to parse examples/src/scala/org/pantsbuild/example/build_file_macros/BUILD:
            TypeError("'int' object is not callable",)
```

Reply:
I think that would be a pretty tricky case to handle. Technically people are allowed to overwrite every value that they import, so I'm not even sure it should be an error. 

I'm not sure it's even possible to handle that case on our end, we just have to be very clear that load statements are evaluated first, and not necessarily in order.